### ISSUE TITLE
Switch completely to TMDb for show data (away from TheTVDB)

### ISSUE THREAD

**Issue by UweTrottmann** [UweTrottmann/SeriesGuide#762]
https://github.com/UweTrottmann/thetvdb-java/issues/35

If TheTVDB.com really is going forward with making the API paid only (on a per-app or per-user basis as indicated at https://thetvdb.com/api-information), then I will strongly consider switching to TMDb for SeriesGuide.

I currently don't see how the community of contributors will accept having to pay (vs. a maybe more forceful donation/non-profit model) for using the data they helped enter in community tools (like Kodi or SeriesGuide).

A switch will obviously create many problems (which is why this was never attempted in the past):
- inconsistent data might lead to loss of marked episodes,
- translations might get removed if TMDb does not have them,
- some shows might no longer be available (I'm pretty sure the TheTVDB catalogue is still larger, esp. with obscure shows).

The alternative would mean either
- make the subscription mandatory to give a share to TheTVDB
- require the user to subscribe at TheTVDB.com and provide an API key or whatever.

**Comment by Cologler** [comment]
How about adding a CDN between SeriesGuide and TheTVDB?

**Comment by UweTrottmann** [comment]
[Trakt is moving away from TheTVDB](https://forums.trakt.tv/t/tvdb-changes-to-a-subscribe-model/3631?u=aeonmckay), that decides it.

# Basic plan
- Create a new database schema that
  - uses auto-assigned IDs instead of TVDB ids for shows, seasons and episodes,
  - moves TVDB id to new property,
  - adds (empty) TMDB id property.

- Switch internals to use the new auto-assigned IDs.

- Switch add/update jobs to use TMDB, including image caching.

- Add a migrater that tries to find out TMDB ids for shows and episodes (required for seasons/groups?).
  - Once TMDB id known for show and all episodes turn on updating (now from TMDB).
  - Keep existing TVDB-only shows, but no longer update them.

- Think about migrating Cloud data (shows keyed by TVDB id, episodes by show TVDB id + season and episode number).

## Things to keep in mind:
- Should support episode groups (e.g. episodes changing group/season).

**Comment by flabbamann** [comment]
Will adding TVDB and/or IMDB IDs to TheMovieDB help for the transition? Those seem to miss on a lot of shows and episodes.

**Comment by UweTrottmann** [comment]
@flabbamann The plan is to let SeriesGuide look up shows via their TVDB ID to find out their TMDb ID, so yes! For episodes TVDB IDs probably won't matter, I guess just letting the code replace by season and episode number is best (also as you said I doubt many have them).

**Comment by chrisnkrueger** [comment]
@UweTrottmann Hey Uwe, I have been working on this problem for years. My app Moviebase (https://play.google.com/store/apps/details?id=com.moviebase) is based on TMDB with a Trak integration.

The app uses and displays the data from TMDB and complement some extra data like episode runtime or aired episodes from Trakt API. The ID matching is solved mostly by TVDB ID and IMDB ID (OMDB API), which works in most cases. In the future, TMDB will go more into the TV show details (see [Trello Episode Runtimes](https://trello.com/c/JTXGkswM/29-add-per-episode-runtimes)).

The biggest issue is the difference data between Trakt/TVDB and TMDB that you already figured out. Problems occur when syncing the user data with Trakt  (especially progress) and mark an episode as watched. 

- Sometimes the TMDB/TVDB IDs are missing on Trakt or vice versa
- Episodes are structured differently (Naruto) or everything in one season (Dragonball).
- Episode numbers are in abs (TMDB) and TVDB not (like One Piece).
- TVDB doesn't split up episodes into two parts. For example House Season 6 Episode 22
https://trakt.tv/shows/house/seasons/6
https://www.themoviedb.org/tv/1408-house/season/6
You can check out my task https://trello.com/c/1KRf2QcP/39-improve-trakt-tvdb-integration-for-tv-shows.

In most cases I have found out, the TMDB episode groups don't reflect the TVDB content correctly. 

Let me know if you need further information.

**Comment by UweTrottmann** [comment]
@chrisnkrueger Thanks, I also think that the app will not support episode groups for the time being and use the traditional season format for the time being. Exactly for the issues you mentioned. Another issue is that TMDB has separate show entries that TVDB recently grouped together (e.g. Midnight Diner: Tokyo Stories is now an additional season of Midnight Diner on TVDB). But these are just a few hiccups that have to be accepted for continued free access to series data.

**Comment by chrisnkrueger** [comment]
@UweTrottmann I am sorry, I didn't think about the link. But you are true, that's a few hiccups.

Episode group doesn't help in every case. For example, in One Piece there is one episode in the middle which is moved to special (TVDB side), and the episode TMDB IDs on the Trakt side don't fit the TMDB episodes. For fixing this issue, I checked the episode number in abs on the server-side.
As ID for episodes, I used the fields showId, seasonNumber and episodeNumber from TMDB.
